### PR TITLE
Merge TS context isolation branch onto Typescript master branch

### PR DESCRIPTION
**Pull request by KiranNiranjan**
Merge `Typescript context isolation` branch onto `Typescript master` branch

### Unit test result
Screenshot 2019-03-19 at 2 42 09 PM

### CODE REVIEW

File: demo/index.html [hunk 1]
```diff
@@ -132,7 +132,9 @@ <h1>Symphony Electron API Demo</h1>
 
             num++;
 
-            var notf = new ssf.Notification(title, {
```
Comment (VishwasShashidhar) [inline demo/index.html]: Can we retain this till we completely move to context isolation?

File: demo/index.html [hunk 2]
```diff
@@ -142,10 +144,13 @@ <h1>Symphony Electron API Demo</h1>
...
-                company: company
-            });
+                company: company,
+                method: 'notification',
+            };
+
+            window.postMessage({ method: 'notification', data: notf }, '*');
 
-            notf.addEventListener('click', onclick);
+            /*notf.addEventListener('click', onclick);
```
Comment (VishwasShashidhar) [inline demo/index.html]: remove commented code

File: src/app/window-handler.ts [hunk 3]
```diff
@@ -285,16 +286,19 @@ export class WindowHandler {
...
         switch (windowType) {
             case 'screen-picker':
-                if (this.screenPickerWindow && this.windowExists(this.screenPickerWindow)) this.screenPickerWindow.close();
+                if (this.screenPickerWindow && windowExists(this.screenPickerWindow)) this.screenPickerWindow.close();
                 break;
             case 'screen-sharing-indicator':
-                if (this.screenSharingIndicatorWindow
-                    && this.windowExists(this.screenSharingIndicatorWindow)) this.screenSharingIndicatorWindow.close();
+                if (winKey) {
+                    const browserWindow = this.windows[ winKey ];
```
Comment (VishwasShashidhar) [inline src/app/window-handler.ts]: remove extra spaces inside the array

File: src/app/window-handler.ts [hunk 4]
```diff
@@ -476,6 +436,55 @@ export class WindowHandler {
...
+        displayId: string,
+        id: number,
+        streamId,
+    ): void {
+        const indicatorScreen =
+            (displayId && electron.screen.getAllDisplays().filter((d) =>
+                displayId.includes(d.id.toString()))[ 0 ]) || electron.screen.getPrimaryDisplay();
+
+        const screenRect = indicatorScreen.workArea;
+        let opts = { ...WindowHandler.getScreenSharingIndicatorOpts(), ...{ winKey: streamId } };
```
Comment (VishwasShashidhar) [inline src/app/window-handler.ts]: can you add a comment above explaining what this logic is about

File: src/common/animation-queue.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            setTimeout(this.animate.bind(this, object), 0);
+        }
+    }
+
+    /**
+     * Animates an animation that is part of the queue
+     * @param object
+     */
+    public animate(object) {
+        object.func.apply(null, object.args)
```
Comment (VishwasShashidhar) [inline src/common/animation-queue.ts]: can we not use async await?

File: src/renderer/notification-handler.ts [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    /**
+     * Sets the position of the notification window
+     *
+     * @param window {BrowserWindow}
+     * @param x {number}
+     * @param y {number}
+     */
+    public setWindowPosition(window: Electron.BrowserWindow, x: number, y: number) {
+        if (window && !window.isDestroyed()) {
+            window.setPosition(parseInt(String(x), 10), parseInt(String(y), 10));
```
Comment (VishwasShashidhar) [inline src/renderer/notification-handler.ts]: what happens if parseInt fails? wouldn't it be better to have defaults?

File: src/renderer/notification-handler.ts [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+        }
+    }
+
+    /**
+     * Initializes / resets the notification positional values
+     */
+    public setupNotificationPosition() {
+        // This feature only applies to windows
+        if (!isMac) {
```
Comment (VishwasShashidhar) [inline src/renderer/notification-handler.ts]: let's invert the condition. If it is mac, return.

File: src/renderer/notification.ts [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+    private readonly activeNotifications: Electron.BrowserWindow[] = [];
+    private readonly inactiveWindows: Electron.BrowserWindow[] = [];
+    private readonly notificationQueue: INotificationData[] = [];
+    private readonly notificationCallbacks: any[] = [];
+    private cleanUpTimer: NodeJS.Timer;
+
+    constructor(opts) {
+        super(opts);
+        ipcMain.on('close-notification', (_event, windowId) => {
+            this.hideNotification(windowId);
```
Comment (VishwasShashidhar) [inline src/renderer/notification.ts]: does this destroy the browser window or just hide it?

Comment (KiranNiranjan) [reply src/renderer/notification.ts]: This just hides the browser window

File: src/app/window-handler.ts (hunk 3 above)
Comment (KiranNiranjan) [reply src/app/window-handler.ts]: This is `TS` recommended

File: demo/index.html (hunk 1 above)
Comment (KiranNiranjan) [reply demo/index.html]: No. In this branch, `ssf` does not exist

File: src/common/api-interface.ts [hunk 9]
```diff
@@ -80,10 +80,28 @@ export interface IBoundsChange extends Electron.Rectangle {
...
+}
+
+export interface ILogMsg {
+    level: LogLevel;
+    details: any;
+    showInConsole: boolean;
+    startTime: number;
+}
+
+export type LogLevel = 'error' | 'warn' | 'info' | 'verbose' | 'debug' | 'silly';
```
Comment (vrbsm) [inline src/common/api-interface.ts]: could you insert a newline at end file ?

File: src/renderer/notification.ts (hunk 8 above)
Comment (VishwasShashidhar) [reply src/renderer/notification.ts]: But won't that be waste of resources. What happens if we close it?

Comment (KiranNiranjan) [reply src/renderer/notification.ts]: Approx 200mb will be reserved based on the display width for about a min. Without this, every notification will be rendered a fresh which looks delayed

File: src/app/window-utils.ts [hunk 10]
```diff
@@ -327,3 +331,11 @@ export const injectStyles = async (mainWindow: BrowserWindow, isCustomTitleBarAn
...
     return await readAndInsertCSS(mainWindow, paths);
 };
+
+/**
+ * Checks if window is valid and exists
+ *
+ * @param window {BrowserWindow}
+ * @return boolean
+ */
+export const windowExists = (window: BrowserWindow): boolean => !!window && typeof window.isDestroyed === 'function' && !window.isDestroyed();
```
Comment (vrbsm) [inline src/app/window-utils.ts]: could you insert a newline at end file ?

File: src/renderer/app-bridge.ts [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+     * Method that broadcast messages to a specific origin via postMessage
+     *
+     * @param method {string}
+     * @param data {any}
+     */
+    private broadcastMessage(method: string, data: any): void {
+        window.postMessage({ method, data }, this.origin);
+    }
+
+}
```
Comment (vrbsm) [inline src/renderer/app-bridge.ts]: could you insert a newline at end file ?

File: src/renderer/components/notification-comp.tsx [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    /**
+     * Sets the About app state
+     *
+     * @param _event
+     * @param data {Object}
+     */
+    private updateState(_event, data): void {
+        this.setState(data as IState);
+    }
+}
```
Comment (vrbsm) [inline src/renderer/components/notification-comp.tsx]: could you insert a newline at end file ?

File: src/renderer/notification-handler.ts [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+            x: firstPosX,
+            y: firstPosY,
+        };
+
+        // Set nextInsertPos
+        this.nextInsertPos.x = this.settings.firstPos.x;
+        this.nextInsertPos.y = this.settings.firstPos.y;
+    }
+
+}
```
Comment (vrbsm) [inline src/renderer/notification-handler.ts]: could you insert a newline at end file ?

File: src/renderer/styles/notification-comp.less [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+  display: flex;
+  align-items: center;
+}
+
+.logo {
+  margin-left: 5px;
+  opacity: 0.6;
+  width: 43px;
+  content: var(--logo-bg);
+}
```
Comment (vrbsm) [inline src/renderer/styles/notification-comp.less]: could you insert a newline at end file ?

File: src/renderer/notification.ts (hunk 8 above)
Comment (VishwasShashidhar) [reply src/renderer/notification.ts]: Understood, thanks for the explanation.

**Comment by KiranNiranjan** [comment]
@VishwasShashidhar @vrbsm I've updated PR as per the changes. Please review. Thanks 🙂

**Pull request merged**